### ISSUE TITLE
eq3bt - error on start, no notification again

### ISSUE THREAD

**Issue by malcb** [fhempy/fhempy#338]
I'm running fhem on raspberry pi without bluetooth so I've used a bluetooth dongle.  This was a small one and I've now got a bigger one with aerials.  With the small one, initially I had the no notifications issue I reported before.  This was fixed with the update although I see the log still reports errors at the start:

This is "small" dongle log:

```
==========
2024-02-10 18:50:26,379 - INFO     - fhempy.lib.fhem_pythonbinding: Starting fhempy 0.1.719...
2024-02-10 18:50:26,400 - INFO     - fhempy.lib.fhem_pythonbinding: Waiting for FHEM connection
2024-02-10 18:50:26,947 - INFO     - websockets.server: server listening on 0.0.0.0:15733
2024-02-10 18:50:35,425 - INFO     - websockets.server: connection open
2024-02-10 18:50:35,433 - INFO     - fhempy.lib.fhem_pythonbinding: Incoming FHEM connection: 127.0.0.1
/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py:225: FutureWarning: BLEDevice.rssi is deprecated and will be removed in a future version of Bleak, use AdvertisementData.rssi instead
  self._dev_hash, "rssi", self._device.rssi, 1
2024-02-10 18:51:03,034 - ERROR    - asyncio: Task exception was never retrieved
future: <Task finished name='Task-286' coro=<eq3bt.notification_received() done, defined at /opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py:197> exception=RuntimeError('dictionary changed size during iteration')>
Traceback (most recent call last):
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py", line 200, in notification_received
    await self.update_all_readings()
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py", line 211, in update_all_readings
    await self.update_schedule_readings()
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py", line 311, in update_schedule_readings
    for day in self.thermostat.schedule.keys():
RuntimeError: dictionary changed size during iteration
2024-02-10 18:51:16,187 - ERROR    - Eq3bt1: Timeout on update, retry in 60s
/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py:225: FutureWarning: BLEDevice.rssi is deprecated and will be removed in a future version of Bleak, use AdvertisementData.rssi instead
  self._dev_hash, "rssi", self._device.rssi, 1
2024-02-10 18:52:24,863 - ERROR    - asyncio: Task exception was never retrieved
future: <Task finished name='Task-743' coro=<eq3bt.notification_received() done, defined at /opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py:197> exception=RuntimeError('dictionary changed size during iteration')>
Traceback (most recent call last):
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py", line 200, in notification_received
    await self.update_all_readings()
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py", line 211, in update_all_readings
    await self.update_schedule_readings()
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/eq3bt/eq3bt.py", line 311, in update_schedule_readings
    for day in self.thermostat.schedule.keys():
RuntimeError: dictionary changed size during iteration
2024-02-10 18:53:03,140 - ERROR    - root: Eq3bt2: readingsBeginUpdate couldn't acquire lock, caused by readingsBeginUpdate without End or Single update inbetween
2024-02-10 18:54:24,968 - ERROR    - root: Eq3bt1: readingsBeginUpdate couldn't acquire lock, caused by readingsBeginUpdate without End or Single update inbetween
2024-02-10 18:57:49,013 - ERROR    - Eq3bt1: Device disconnected, reconnect now
2024-02-10 18:57:49,133 - ERROR    - Eq3bt2: Device disconnected, reconnect now
==========
```

With aerial dongle I always have "connected (no notifications)" no matter how long I wait.  I've gone through the bluetoothctl steps as in the eq3bt page.  That all works and shows no errors.  Both dongles report as Realtek but different mac addresses.  I have 2 eq3bt thermostats, Wq3bt1 and Eq3bt2.  The log for aerial dongle is:

```
==========
Activating virtual environment...OK
2024-02-14 12:48:58,038 - INFO     - fhempy.lib.fhem_pythonbinding: Starting fhempy 0.1.719...
2024-02-14 12:48:58,059 - INFO     - fhempy.lib.fhem_pythonbinding: Waiting for FHEM connection
2024-02-14 12:48:58,522 - INFO     - websockets.server: server listening on 0.0.0.0:15733
2024-02-14 12:49:07,854 - INFO     - websockets.server: connection open
2024-02-14 12:49:07,862 - INFO     - fhempy.lib.fhem_pythonbinding: Incoming FHEM connection: 127.0.0.1
2024-02-14 12:50:07,527 - ERROR    - Eq3bt2: Timeout on update, retry in 60s
2024-02-14 12:51:37,574 - ERROR    - Eq3bt2: Timeout on update, retry in 60s
/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py:225: FutureWarning: BLEDevice.rssi is deprecated and will be removed in a future version of Bleak, use AdvertisementData.rssi instead
  self._dev_hash, "rssi", self._device.rssi, 1
/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py:225: FutureWarning: BLEDevice.rssi is deprecated and will be removed in a future version of Bleak, use AdvertisementData.rssi instead
  self._dev_hash, "rssi", self._device.rssi, 1
2024-02-14 12:52:44,786 - ERROR    - Eq3bt2: Failed to subscribe to notifications
Traceback (most recent call last):
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py", line 243, in connect_once
    await self._subscribe_notifies()
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py", line 276, in _subscribe_notifies
    await self._client.start_notify(
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/bleak/__init__.py", line 717, in start_notify
    await self._backend.start_notify(characteristic, wrapped_callback, **kwargs)
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/bleak/backends/bluezdbus/client.py", line 1003, in start_notify
    assert_reply(reply)
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/bleak/backends/bluezdbus/utils.py", line 22, in assert_reply
    raise BleakDBusError(reply.error_name, reply.body)
bleak.exc.BleakDBusError: [org.bluez.Error.NotPermitted] Not paired
2024-02-14 12:52:46,774 - ERROR    - Eq3bt1: Failed to subscribe to notifications
Traceback (most recent call last):
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py", line 243, in connect_once
    await self._subscribe_notifies()
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/fhempy/lib/core/bluetoothle.py", line 276, in _subscribe_notifies
    await self._client.start_notify(
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/bleak/__init__.py", line 717, in start_notify
    await self._backend.start_notify(characteristic, wrapped_callback, **kwargs)
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/bleak/backends/bluezdbus/client.py", line 1003, in start_notify
    assert_reply(reply)
  File "/opt/fhem/.fhempy/fhempy_venv/lib/python3.9/site-packages/bleak/backends/bluezdbus/utils.py", line 22, in assert_reply
    raise BleakDBusError(reply.error_name, reply.body)
bleak.exc.BleakDBusError: [org.bluez.Error.NotPermitted] Not paired
=========

```
What should I try to fix this?  When I swap back to small dongle (power off first of course) the notifications work again.

**Comment by fhempy** [comment]
Can you please provide the output of
```
bluetoothctl
paired-devices
devices Paired
list
info MAC
```

**Comment by malcb** [comment]
SSH'ed in to pi and started bluetoothctl and I get
```

[CC-RT-BLE]# devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
Device 00:1A:22:1B:D3:05 CC-RT-BLE

[CC-RT-BLE]# list
Controller 8C:88:4B:07:12:97 pi [default]

[CC-RT-BLE]# info 00:1A:22:1B:D1:DA
Device 00:1A:22:1B:D1:DA (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: no
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........
[CC-RT-BLE]# info 00:1A:22:1B:D3:05
Device 00:1A:22:1B:D3:05 (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: no
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00  
```        
So neither thermostats are paired (paired-devices returns nothing).  I now pair these in bluetoothctl.

```
[CC-RT-BLE]# pair 00:1A:22:1B:D1:DA
Attempting to pair with 00:1A:22:1B:D1:DA
[CHG] Device 00:1A:22:1B:D1:DA Paired: yes
Pairing successful
[CC-RT-BLE]# paired-devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
[CC-RT-BLE]# pair 00:1A:22:1B:D3:05
Attempting to pair with 00:1A:22:1B:D3:05
[CHG] Device 00:1A:22:1B:D3:05 Paired: yes
Pairing successful
[CC-RT-BLE]# paired-devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
Device 00:1A:22:1B:D3:05 CC-RT-BLE
[CC-RT-BLE]#

[CC-RT-BLE]# info 00:1A:22:1B:D3:05
Device 00:1A:22:1B:D3:05 (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: yes
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........
[CC-RT-BLE]# info 00:1A:22:1B:D1:DA
Device 00:1A:22:1B:D1:DA (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: yes
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........
[CC-RT-BLE]#
```

So it all seems to work in bluetoothctl.  I have not uninstalled/reinstalled bluetooth.  Swapping the dongles seems to work without doing that.

**Comment by malcb** [comment]
BTW tried the update to 0.1.720.  No difference.  I did notice after the update that 1 thermostat connects and I can change the setting.  The second is still connecting.  After a wait that one is also connected and I can change that setting.  Both are not getting notification though.  Tried a fhem restart and no difference.

**Comment by fhempy** [comment]
I can see that you have successfully paired both devices. Can you please check if the paired-devices are still there after a fhempy restart?

**Comment by malcb** [comment]
No.  Here is the output of bluetoothctl.  pi is just powered on and booted.  fhem is booting.

```
root@pi:~# bluetoothctl
Agent registered
[CC-RT-BLE]# list
Controller 8C:88:4B:07:12:97 pi [default]
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: yes
[CC-RT-BLE]# pair  00:1A:22:1B:D3:05
Attempting to pair with 00:1A:22:1B:D3:05
Failed to pair: org.bluez.Error.InProgress
[CC-RT-BLE]# info  00:1A:22:1B:D3:05
Device 00:1A:22:1B:D3:05 (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: no
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........
[CC-RT-BLE]# pair  00:1A:22:1B:D3:05
Attempting to pair with 00:1A:22:1B:D3:05
Failed to pair: org.bluez.Error.InProgress
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Device 00:1A:22:1B:D1:DA RSSI: -68
[CHG] Device 00:1A:22:1B:D1:DA RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Device 00:1A:22:1B:D3:05 RSSI: -66
[CHG] Device 00:1A:22:1B:D3:05 RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D1:DA Connected: yes
[CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: yes
[CHG] Device 00:1A:22:1B:D3:05 Connected: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: yes
[CC-RT-BLE]# info  00:1A:22:1B:D3:05
Device 00:1A:22:1B:D3:05 (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: no
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........
[CC-RT-BLE]# pair  00:1A:22:1B:D3:05
Attempting to pair with 00:1A:22:1B:D3:05
[CHG] Device 00:1A:22:1B:D3:05 Paired: yes
Pairing successful
[CC-RT-BLE]# info 00:1A:22:1B:D1:DA
Device 00:1A:22:1B:D1:DA (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: no
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........
[CC-RT-BLE]# pair 00:1A:22:1B:D1:DA
Attempting to pair with 00:1A:22:1B:D1:DA
[CHG] Device 00:1A:22:1B:D1:DA Paired: yes
Pairing successful
[CC-RT-BLE]# info 00:1A:22:1B:D1:DA
Device 00:1A:22:1B:D1:DA (public)
        Name: CC-RT-BLE
        Alias: CC-RT-BLE
        Paired: yes
        Trusted: yes
        Blocked: no
        Connected: yes
        LegacyPairing: no
        UUID: Generic Access Profile    (00001800-0000-1000-8000-00805f9b34fb)
        UUID: Generic Attribute Profile (00001801-0000-1000-8000-00805f9b34fb)
        UUID: Device Information        (0000180a-0000-1000-8000-00805f9b34fb)
        UUID: Vendor specific           (3e135142-654f-9090-134a-a6ff5bb77046)
        UUID: Vendor specific           (9e5d1e47-5c13-43a0-8635-82ad38a1386f)
        ManufacturerData Key: 0x0000
        ManufacturerData Value:
  00 00 00 00 00 00 00 00 00                       .........
[CC-RT-BLE]#
```
maybe this is useful?

```
[CC-RT-BLE]# export
SHELL=0xd73858
HISTORY=0x4a1f8
DBUS_CONNECTION=0xd92ae8
/org/bluez/hci0=0xd979c8
/org/bluez/hci0/dev_00_1A_22_1B_D1_DA=0xd998e8
/org/bluez/hci0/dev_00_1A_22_1B_D3_05=0xda4a88
[CC-RT-BLE]# paired-devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
Device 00:1A:22:1B:D3:05 CC-RT-BLE
[CC-RT-BLE]#
```
update to 0.1.721, restart fhempy, ~10min wait, still connected (no notifications)

```
[CC-RT-BLE]# paired-devices
[CC-RT-BLE]# pair 00:1A:22:1B:D1:DA
Attempting to pair with 00:1A:22:1B:D1:DA
[CHG] Device 00:1A:22:1B:D1:DA Paired: yes
Pairing successful
[CC-RT-BLE]# pair  00:1A:22:1B:D3:05
Attempting to pair with 00:1A:22:1B:D3:05
[CHG] Device 00:1A:22:1B:D3:05 Paired: yes
Pairing successful
[CC-RT-BLE]# paired-devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
Device 00:1A:22:1B:D3:05 CC-RT-BLE
[CC-RT-BLE]#
```

reboot fhem

```
[CC-RT-BLE]# paired-devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
Device 00:1A:22:1B:D3:05 CC-RT-BLE
[CC-RT-BLE]# paired-devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
Device 00:1A:22:1B:D3:05 CC-RT-BLE
[CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: no
[CHG] Device 00:1A:22:1B:D1:DA Paired: no
[CHG] Device 00:1A:22:1B:D1:DA Connected: no
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: no
[CHG] Device 00:1A:22:1B:D3:05 Paired: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: no
[bluetooth]# paired-devices
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Device 00:1A:22:1B:D1:DA RSSI: -72
[CHG] Device 00:1A:22:1B:D3:05 RSSI: -66
[NEW] Device 55:F3:28:8F:61:C1 55-F3-28-8F-61-C1
[CHG] Device 55:F3:28:8F:61:C1 TxPower is nil
[CHG] Device 55:F3:28:8F:61:C1 RSSI is nil
[CHG] Device 00:1A:22:1B:D3:05 RSSI is nil
[CHG] Device 00:1A:22:1B:D1:DA RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: yes
[DEL] Device 55:F3:28:8F:61:C1 55-F3-28-8F-61-C1
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Device 00:1A:22:1B:D1:DA RSSI: -66
[CHG] Device 00:1A:22:1B:D1:DA RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Device 00:1A:22:1B:D3:05 RSSI: -66
[CHG] Device 00:1A:22:1B:D1:DA RSSI: -60
[CHG] Device 00:1A:22:1B:D1:DA RSSI is nil
[CHG] Device 00:1A:22:1B:D3:05 RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D1:DA Connected: yes
[CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: yes
[CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: no
[CHG] Device 00:1A:22:1B:D1:DA Connected: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D1:DA Connected: yes
[CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: yes
[CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: no
[CHG] Device 00:1A:22:1B:D1:DA Connected: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[NEW] Device 55:F3:28:8F:61:C1 55-F3-28-8F-61-C1
[NEW] Device 5F:9D:18:CD:FF:79 5F-9D-18-CD-FF-79
[CHG] Device 00:1A:22:1B:D3:05 RSSI: -68
[CHG] Device 00:1A:22:1B:D3:05 RSSI is nil
[CHG] Device 5F:9D:18:CD:FF:79 RSSI is nil
[CHG] Device 55:F3:28:8F:61:C1 TxPower is nil
[CHG] Device 55:F3:28:8F:61:C1 RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Device 55:F3:28:8F:61:C1 RSSI: -100
[CHG] Device 55:F3:28:8F:61:C1 TxPower: 11
[NEW] Device 45:21:CA:18:EF:4B 45-21-CA-18-EF-4B
[CHG] Device 00:1A:22:1B:D3:05 RSSI: -68
[CHG] Device 00:1A:22:1B:D3:05 Connected: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: yes
[DEL] Device 5F:9D:18:CD:FF:79 5F-9D-18-CD-FF-79
[CHG] Device 00:1A:22:1B:D3:05 RSSI is nil
[CHG] Device 45:21:CA:18:EF:4B TxPower is nil
[CHG] Device 45:21:CA:18:EF:4B RSSI is nil
[CHG] Device 55:F3:28:8F:61:C1 TxPower is nil
[CHG] Device 55:F3:28:8F:61:C1 RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[DEL] Device 55:F3:28:8F:61:C1 55-F3-28-8F-61-C1
[DEL] Device 45:21:CA:18:EF:4B 45-21-CA-18-EF-4B
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[NEW] Device 55:F3:28:8F:61:C1 55-F3-28-8F-61-C1
[NEW] Device 4C:02:A7:CF:0D:48 4C-02-A7-CF-0D-48
[CHG] Device 00:1A:22:1B:D1:DA RSSI: -72
[CHG] Device 00:1A:22:1B:D1:DA RSSI is nil
[CHG] Device 4C:02:A7:CF:0D:48 TxPower is nil
[CHG] Device 4C:02:A7:CF:0D:48 RSSI is nil
[CHG] Device 55:F3:28:8F:61:C1 TxPower is nil
[CHG] Device 55:F3:28:8F:61:C1 RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D1:DA Connected: yes
[CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: yes
[CC-RT-BLE]#
```

checked if still paired,  they are not, so pair and recheck.

```
[CC-RT-BLE]# paired-devices
[CC-RT-BLE]# pair 00:1A:22:1B:D1:DA
Attempting to pair with 00:1A:22:1B:D1:DA
[CHG] Device 00:1A:22:1B:D1:DA Paired: yes
Pairing successful
[CC-RT-BLE]# pair  00:1A:22:1B:D3:05
Attempting to pair with 00:1A:22:1B:D3:05
[CHG] Device 00:1A:22:1B:D3:05 Paired: yes
Pairing successful
[CC-RT-BLE]# paired-devices
Device 00:1A:22:1B:D1:DA CC-RT-BLE
Device 00:1A:22:1B:D3:05 CC-RT-BLE
[CC-RT-BLE]#

```

So for some reason fhempy drops the pairing and then fails to pair.  Not I don't have the keep_connected attribute so that will be at whatever is the default, "no" I assume.

**Comment by fhempy** [comment]
Please try 
```
bluetoothctl
remove 00:1A:22:1B:D1:DA
remove 00:1A:22:1B:D3:05
```
After that restart fhempy, wait a few minutes and check in bluetoothctl if the devices are paired.

**Comment by malcb** [comment]
That might have helped but it not cured the problem.  Eventually one thermostat has connected but one hasn't.  I wonder if the problem is that the large dongle is more sensitive than the small one and it is picking up bluetooth kit from next door (or maybe mine in another room, but I don't recognise it as mine).  With bluetoothctl running I see a lot of instances like this:

```
[CHG] Controller 8C:88:4B:07:12:97 Discovering: yes
[CHG] Device 00:1A:22:1B:D3:05 RSSI: -42
[NEW] Device 7A:7F:50:31:DB:61 7A-7F-50-31-DB-61
[CHG] Device 7A:7F:50:31:DB:61 TxPower is nil
[CHG] Device 7A:7F:50:31:DB:61 RSSI is nil
[CHG] Device 00:1A:22:1B:D3:05 RSSI is nil
[CHG] Controller 8C:88:4B:07:12:97 Discovering: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: yes
[CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: no
[CHG] Device 00:1A:22:1B:D3:05 Connected: no
[DEL] Device 7A:7F:50:31:DB:61 7A-7F-50-31-DB-61
```
The new/del pair mac address varies.  I've also caught this 

```
[CHG] Device 00:1A:22:1B:D3:05 Connected: yes
Request passkey
[agent] Enter passkey (number in 0-999999): [CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: yes
[agent] Enter passkey (number in 0-999999): [CHG] Device 00:1A:22:1B:D1:DA Connected: yes
[agent] Enter passkey (number in 0-999999): [CHG] Device 00:1A:22:1B:D1:DA ServicesResolved: yes
[agent] Enter passkey (number in 0-999999): [DEL] Device 44:C2:AC:A7:83:9B 44-C2-AC-A7-83-9B
[agent] Enter passkey (number in 0-999999): [CHG] Device 00:1A:22:1B:D3:05 ServicesResolved: no
[bluetooth]# [CHG] Device 00:1A:22:1B:D3:05 Connected: no
[bluetooth]# Request canceled
```

At this time ...DA is fully connected, ...05 is "connecting" and have been for a long time.

I've tried remove on both parts again but the end results in DA becomes fully connected again and 05 is still connecting.

**Comment by malcb** [comment]
Here is fhempy log after a a fhem shutdown restart

```
Activating virtual environment...OK
2024-02-16 14:11:23,170 - INFO     - fhempy.lib.fhem_pythonbinding: Starting fhempy 0.1.721...
2024-02-16 14:11:23,190 - INFO     - fhempy.lib.fhem_pythonbinding: Waiting for FHEM connection
2024-02-16 14:11:23,658 - INFO     - websockets.server: server listening on 0.0.0.0:15733
2024-02-16 14:11:33,081 - INFO     - websockets.server: connection open
2024-02-16 14:11:33,089 - INFO     - fhempy.lib.fhem_pythonbinding: Incoming FHEM connection: 127.0.0.1
2024-02-16 14:12:10,660 - ERROR    - Eq3bt2: Timeout on update, retry in 60s
2024-02-16 14:12:29,561 - ERROR    - Eq3bt1: Timeout exceeded.
<pexpect.pty_spawn.spawn object at 0xb32588e0>
command: /usr/bin/bluetoothctl
args: [b'/usr/bin/bluetoothctl']
buffer (last 100 chars): 'ndor specific\r\nPairing successful\r\n[\x01\x1b[0;91m\x02DEL\x01\x1b[0m\x02] Device 76:F8:69:2E:2F:E6 76-F8-69-2E-2F-E6\r\n'
before (last 100 chars): 'ndor specific\r\nPairing successful\r\n[\x01\x1b[0;91m\x02DEL\x01\x1b[0m\x02] Device 76:F8:69:2E:2F:E6 76-F8-69-2E-2F-E6\r\n'
after: <class 'pexpect.exceptions.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 1714
child_fd: 14
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('#')
    1: EOF
2024-02-16 14:13:04,667 - ERROR    - Eq3bt1: Timeout exceeded.
<pexpect.pty_spawn.spawn object at 0xb32588e0>
command: /usr/bin/bluetoothctl
args: [b'/usr/bin/bluetoothctl']
buffer (last 100 chars): 'F8-69-2E-2F-E6\r\n\x1b[1;39m[agent] Enter passkey (number in 0-999999): \x1b[0mChanging power on succeeded\r\n'
before (last 100 chars): 'F8-69-2E-2F-E6\r\n\x1b[1;39m[agent] Enter passkey (number in 0-999999): \x1b[0mChanging power on succeeded\r\n'
after: <class 'pexpect.exceptions.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 1714
child_fd: 14
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('#')
    1: EOF
2024-02-16 14:13:34,738 - ERROR    - Eq3bt1: Timeout exceeded.
<pexpect.pty_spawn.spawn object at 0xb32588e0>
command: /usr/bin/bluetoothctl
args: [b'/usr/bin/bluetoothctl']
buffer (last 100 chars): '0:1A:22:1B:D3:05 ServicesResolved: no\r\n[\x01\x1b[0;93m\x02CHG\x01\x1b[0m\x02] Device 00:1A:22:1B:D3:05 Connected: no\r\n'
before (last 100 chars): '0:1A:22:1B:D3:05 ServicesResolved: no\r\n[\x01\x1b[0;93m\x02CHG\x01\x1b[0m\x02] Device 00:1A:22:1B:D3:05 Connected: no\r\n'
after: <class 'pexpect.exceptions.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 1714
child_fd: 14
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('#')
    1: EOF
2024-02-16 14:13:40,709 - ERROR    - Eq3bt2: Timeout on update, retry in 60s
2024-02-16 14:14:04,839 - ERROR    - Eq3bt1: Timeout exceeded.
<pexpect.pty_spawn.spawn object at 0xb32588e0>
command: /usr/bin/bluetoothctl
args: [b'/usr/bin/bluetoothctl']
buffer (last 100 chars): 'ected: no\r\n\x1b[1;39m[agent] Enter passkey (number in 0-999999): \x1b[0mDefault agent request successful\r\n'
before (last 100 chars): 'ected: no\r\n\x1b[1;39m[agent] Enter passkey (number in 0-999999): \x1b[0mDefault agent request successful\r\n'
after: <class 'pexpect.exceptions.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 1714
child_fd: 14
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('#')
    1: EOF
2024-02-16 14:14:34,929 - ERROR    - Eq3bt1: Timeout exceeded.
<pexpect.pty_spawn.spawn object at 0xb32588e0>
command: /usr/bin/bluetoothctl
args: [b'/usr/bin/bluetoothctl']
buffer (last 100 chars): 'evice 00:1A:22:1B:D3:05 RSSI: -66\r\n[\x01\x1b[0;92m\x02NEW\x01\x1b[0m\x02] Device 73:10:B0:7A:A8:30 73-10-B0-7A-A8-30\r\n'
before (last 100 chars): 'evice 00:1A:22:1B:D3:05 RSSI: -66\r\n[\x01\x1b[0;92m\x02NEW\x01\x1b[0m\x02] Device 73:10:B0:7A:A8:30 73-10-B0-7A-A8-30\r\n'
after: <class 'pexpect.exceptions.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 1714
child_fd: 14
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('#')
    1: EOF
2024-02-16 14:15:10,760 - ERROR    - Eq3bt2: Timeout on update, retry in 60s
2024-02-16 14:15:15,016 - ERROR    - Eq3bt1: Timeout exceeded.
<pexpect.pty_spawn.spawn object at 0xb32588e0>
command: /usr/bin/bluetoothctl
args: [b'/usr/bin/bluetoothctl']
buffer (last 100 chars): ':10:B0:7A:A8:30 73-10-B0-7A-A8-30\r\n[\x01\x1b[0;91m\x02DEL\x01\x1b[0m\x02] Device 76:F8:69:2E:2F:E6 76-F8-69-2E-2F-E6\r\n'
before (last 100 chars): ':10:B0:7A:A8:30 73-10-B0-7A-A8-30\r\n[\x01\x1b[0;91m\x02DEL\x01\x1b[0m\x02] Device 76:F8:69:2E:2F:E6 76-F8-69-2E-2F-E6\r\n'
after: <class 'pexpect.exceptions.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 1714
child_fd: 14
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('#')
    1: EOF
2024-02-16 14:15:49,114 - ERROR    - Eq3bt1: Timeout exceeded.
<pexpect.pty_spawn.spawn object at 0xb32588e0>
command: /usr/bin/bluetoothctl
args: [b'/usr/bin/bluetoothctl']
buffer (last 100 chars): '-43a0-8635-82ad38a1386f\r\n\tVendor specific\r\n[\x01\x1b[0;91m\x02DEL\x01\x1b[0m\x02] Device 00:1A:22:1B:D3:05 CC-RT-BLE\r\n'
before (last 100 chars): '-43a0-8635-82ad38a1386f\r\n\tVendor specific\r\n[\x01\x1b[0;91m\x02DEL\x01\x1b[0m\x02] Device 00:1A:22:1B:D3:05 CC-RT-BLE\r\n'
after: <class 'pexpect.exceptions.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 1714
child_fd: 14
closed: False
timeout: 30
delimiter: <class 'pexpect.exceptions.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
searcher: searcher_re:
    0: re.compile('#')
    1: EOF
```

Why is fhempy trying to connect to 76:F8:69:2E:2F:E6 76-F8-69-2E-2F-E6 ?  Neither part is connected currently, both just say connecting.  I can't even change the set temperature so I'm going to reboot the pi.

**Comment by malcb** [comment]
I've rebooted and I also check the number displayed when the pair button is pushed and that had changed for one.  Why I don't know.  It's not a typo because it was working on the other dongle and the number is way different.  I've changed that def statement (it was the one that would not connect before).  Both are now "connected" and seem to working.  There are some line in the log I'm not sure about.  I'll see what appears in the log at next boot tomorrow and report that, but for now it is working again although not 100% sure why.  Perhaps a combination of the remove and finding the change in key.

**Comment by fhempy** [comment]
As far as I know the pairing key changes when you change the battery, but if the pairing was successfully done once, it keeps the paired connection active.
I'll try to add a reading which informs about the wrong pairing pin.

**Comment by fhempy** [comment]
Please open a new issue if the error occurs again.
